lib/debugger/debugger.dart

```diff
@@ -0,0 +1,1309 @@
+      if (executionPoint.position != null) {
+        _showLineClass(executionPoint.position.line - 1);
+      }
+    }
+  }
+
+  void _clearLineClass() {
+    if (_currentLineClass != null) {
+      codeMirror.removeLineClass(
+          _currentLineClass, 'background', 'executionLine');
+    }
+    _currentLineClass = null;
+
+    _executionPointElement?.dispose();
+    _executionPointElement = null;
+  }
+
+  void _showLineClass(int line) {
+    if (_currentLineClass != null) {
+      _clearLineClass();
+    }
+    _currentLineClass = line;
+    codeMirror.addLineClass(_currentLineClass, 'background', 'executionLine');
+  }
+
+  void displayExecutionPoint(Script script, {Pos position}) {
+    executionPoint = new ScriptAndPos(script.uri, position: position);
+
+    // This also calls _refreshMarkers().
+    displayScript(script, scrollTo: position);
+
+    _executionPointElement?.dispose();
+    _executionPointElement = null;
+
+    if (script.source != null && position != null) {
+      _executionPointElement =
+          span(c: 'octicon octicon-arrow-up execution-marker');
+
+      codeMirror.addWidget(
+        new Position(position.line - 1, position.column - 1),
+        _executionPointElement.element,
+      );
+    }
+  }
+
+  void clearExecutionPoint() {
+    executionPoint = null;
+    _clearLineClass();
+    _refreshMarkers();
+  }
+
+  final Map<String, int> _lastScrollPositions = <String, int>{};
+
+  void displayScript(Script newScript, {Pos scrollTo}) {
+    if (currentScript != null) {
+      final ScrollInfo scrollInfo = codeMirror.getScrollInfo();
+      _lastScrollPositions[currentScript.uri] = scrollInfo.top;
+    }
+
+    final bool sameScript = currentScript?.uri == newScript?.uri;
+
+    currentScript = newScript;
+
+    if (newScript == null) {
+      codeMirror.getDoc().setValue('');
+    } else {
+      // TODO(devoncarew): set the mode to either dart or javascript
+      // codeMirror.setMode(mode);
+
+      if (!sameScript) {
+        final String source = newScript?.source ?? '<source not available>';
+        codeMirror.getDoc().setValue(source);
+      }
+
+      if (scrollTo != null) {
+        codeMirror.scrollIntoView(scrollTo.line - 1, 0, margin: 150);
+      } else {
+        final int top = _lastScrollPositions[newScript.uri] ?? 0;
+        codeMirror.scrollTo(0, top);
+      }
+    }
+
+    _executionPointElement?.dispose();
+    _executionPointElement = null;
+
+    _refreshMarkers();
+  }
+}
+
+typedef URIDescriber = String Function(String uri);
+
+class BreakpointsView {
+  BreakpointsView(this._breakpointsCountDiv, DebuggerState debuggerState,
+      URIDescriber uriDescriber) {
+    _items = new SelectableList<Breakpoint>()
+      ..flex()
+      ..clazz('menu-item-bottom-border')
+      ..element.style.overflowY = 'scroll';
+
+    _items.setRenderer((Breakpoint breakpoint) {
+      final dynamic location = breakpoint.location;
+
+      final CoreElement element = li(text: '', c: 'list-item');
```

Comment:
can text parameter be omitted as it is empty?

Reply:
yup, done